File: Pc191/Bit_manipulation/skyline.py
import heapq

def getSkyline(buildings):
    events = []
    for L, R, H in buildings:
        events.append((L, -H, R))   # entering building
    # Only entering events in heap version
    events.sort()

    result = []
    heap = [(0, float('inf'))]  # (negative height, right x)
    prev_max = 0

    i = 0
    n = len(events)

    # process all x positions where buildings start or end
    points = sorted(set([L for L, _, _ in events] + [R for _, _, R in events]))

    for x in points:
        # add new buildings whose left <= x
        while i < n and events[i][0] <= x:
            L, negH, R = events[i]
            heapq.heappush(heap, (negH, R))
            i += 1

        # remove buildings that have ended before x
        while heap and heap[0][1] <= x:
            heapq.heappop(heap)

        cur_max = -heap[0][0]
        if cur_max != prev_max:
            result.append([x, cur_max])
            prev_max = cur_max

    return result


# Example test
buildings = [[2,9,10],[3,7,15],[5,12,12],[15,20,10],[19,24,8]]
print(getSkyline(buildings))

File: Pc191/Bit_manipulation/skyline.cpp
#include <bits/stdc++.h>
using namespace std;

class Solutuion
{
public:
    vector<vector<int>> getSkyline(vector<vector<int>> &buildings)
    {
        vector<pair<int, int>> events;

        for (auto &b : buildings)
        {
            int l = b[0], r = b[1], h = b[2];
            events.push_back({l, -h});
            events.push_back({r, h});
        }

        sort(events.begin(), events.end(), [](auto &a, auto &b)
             {
                if(a.first==b.first)return a.second<b.second;
                return a.first<b.first; });

        multiset<int> heights = {0};
        int prevMax = 0;
        vector<vector<int>> result;

        for (auto &[x, h] : events)
        {
            if (h < 0)
                heights.insert(-h);
            else
                heights.erase(heights.find(h));

            int cur_max = *heights.rbegin();

            if (curMax != prevMax)
            {
                result.push_back({x, cur_max});
                prevMax = curMax;
            }
        }

        return result;
    }
};

int main()
{

    return 0;
}